### ISSUE TITLE
Couldn't get the endpoint from output of modules/beta-private-cluster if set enable_private_endpoint to true

### ISSUE THREAD

**Issue by FuyuanChu** [terraform-google-modules/terraform-google-kubernetes-engine#1479]
### TL;DR

TL;DR

We are trying to change the GKE cluster provisioned using `beta-private-cluster` to use the private endpoint.
So we set the variable `enable_private_endpoint` to true and set `master_authorized_networks` to `[]`.
After the change, the terraform plan is broken. It seems the provider `kubernetes` couldn't get the correct endpoint using `${module.gke.endpoint}` after we set enable_private_endpoint to true. So  provider `kubernetes` fallback to use the default one `localhost` as the host.

I find a variable `deploy_using_private_endpoint` could change the output value of endpoint, but it doesn't work. The provider couldn't get the `module.gke.endpoint` no matter if it was set to true or false.

### Expected behavior

The provider `kubernetes` should get the correct endpoint of the cluster, no matter the public one or the private one.
```
provider "kubernetes" {
  host                   = "https://${module.gke.endpoint}"
  token                  = data.google_client_config.default.access_token
  cluster_ca_certificate = base64decode(module.gke.ca_certificate)
}
```

The plan should success.

I find a variable `deploy_using_private_endpoint` could change the output value of endpoint, but it doesn't work. The provider couldn't get the `module.gke.endpoint` no matter it set to true or false.

### Observed behavior

The provider fallback to use the `localhost` as the host.

```
module.gke.random_string.cluster_service_account_suffix: Refreshing state... [id=ctox]
module.gke.google_compute_firewall.master_webhooks[0]: Refreshing state... [id=xxxxxxxxxxxxxxxxx]
module.gke.random_shuffle.available_zones: Refreshing state... [id=-]
google_storage_bucket_object.updated_subnet_list: Refreshing state... [id=xxxxxxxxxxxxxx]
module.gke.google_container_cluster.primary: Refreshing state... [id=xxxxxxxxxxxxx]
module.gke.google_container_node_pool.pools["default-node-pool"]: Refreshing state... [id=xxxxxxxxxxxx]
kubernetes_storage_class.pd_ssd_storage_class[0]: Refreshing state... [id=xxxxxxxxx]
google_storage_bucket.gitlab_cache_bucket[0]: Refreshing state... [id=xxxxxxxxxxxxxxxx]
kubernetes_namespace.tools: Refreshing state... [id=xxxxxxxxxx]
google_service_account_iam_member.cert_manager[0]: Refreshing state... [id=xxxxxxxxxxxxxxxx]
google_service_account_key.gke_sa_key: Refreshing state... [id=xxxxxxxxxxxxxxx]
google_service_account_key.project_sa_key[0]: Refreshing state... [id=xxxxxxxxxxxx]
google_service_account_iam_binding.gitlab_account_iam: Refreshing state... [id=xxxxxxxxxxxxxxxx]
google_storage_bucket_iam_binding.cache_object_creator[0]: Refreshing state... [id=xxxxxxxxxxxxxxxxx]
time=2022-11-25T06:25:02Z level=error msg=Module /builds/xxxxxxxxxxx has finished with an error: 1 error occurred:
	* exit status 1
 prefix=[/builds/xxxxxxxxxxx] 
time=2022-11-25T06:25:02Z level=info msg=╷
│ Error: Get "http://localhost/api/v1/namespaces/infra": dial tcp 127.0.0.1:80: connect: connection refused
│ 
│   with kubernetes_namespace.tools,
│   on main.tf line 207, in resource "kubernetes_namespace" "tools":
│  207: resource "kubernetes_namespace" "tools" {
│ 
╵
```

### Terraform Configuration
We use the module `beta-private-cluster` just like the example in `README.md`. 

```hcl



# google_client_config and kubernetes provider must be explicitly specified like the following.
data "google_client_config" "default" {}

provider "kubernetes" {
  host                   = "https://${module.gke.endpoint}"    # Here we reference the endpoint of the cluster using the output of module gke
  token                  = data.google_client_config.default.access_token
  cluster_ca_certificate = base64decode(module.gke.ca_certificate)
}

module "gke" {
  source                     = "terraform-google-modules/kubernetes-engine/google//modules/beta-private-cluster"
  project_id                 = "<PROJECT ID>"
  name                       = "gke-test-1"
  region                     = "us-central1"
  zones                      = ["us-central1-a", "us-central1-b", "us-central1-f"]
  network                    = "vpc-01"
  subnetwork                 = "us-central1-01"
  ip_range_pods              = "us-central1-01-gke-01-pods"
  ip_range_services          = "us-central1-01-gke-01-services"
  http_load_balancing        = false
  network_policy             = false
  horizontal_pod_autoscaling = true
  filestore_csi_driver       = false
  enable_private_endpoint    = var.enable_private_endpoint      # var.enable_private_endpoint set to true to enable the private endpoint
  enable_private_nodes       = true
  master_ipv4_cidr_block     = "10.0.0.0/28"
  istio                      = true
  cloudrun                   = true
  dns_cache                  = false
  master_authorized_networks = var.enable_private_endpoint ? [] : concat(var.master_authorized_networks, var.additional_master_authorized_networks)      # remove master_authorized_networks once we enable the private endpoint
  node_pools = [
    {
      name                      = "default-node-pool"
      machine_type              = "e2-medium"
      node_locations            = "us-central1-b,us-central1-c"
      min_count                 = 1
      max_count                 = 100
      local_ssd_count           = 0
      spot                      = false
      local_ssd_ephemeral_count = 0
      disk_size_gb              = 100
      disk_type                 = "pd-standard"
      image_type                = "COS_CONTAINERD"
      enable_gcfs               = false
      enable_gvnic              = false
      auto_repair               = true
      auto_upgrade              = true
      service_account           = "project-service-account@<PROJECT ID>.iam.gserviceaccount.com"
      preemptible               = false
      initial_node_count        = 80
    },
  ]

  node_pools_oauth_scopes = {
    all = [
      "https://www.googleapis.com/auth/logging.write",
      "https://www.googleapis.com/auth/monitoring",
    ]
  }

  node_pools_labels = {
    all = {}

    default-node-pool = {
      default-node-pool = true
    }
  }

  node_pools_metadata = {
    all = {}

    default-node-pool = {
      node-pool-metadata-custom-value = "my-node-pool"
    }
  }

  node_pools_taints = {
    all = []

    default-node-pool = [
      {
        key    = "default-node-pool"
        value  = true
        effect = "PREFER_NO_SCHEDULE"
      },
    ]
  }

  node_pools_tags = {
    all = []

    default-node-pool = [
      "default-node-pool",
    ]
  }
}
```
```


### Terraform Version

```sh
$ terraform --version
Terraform v1.0.9
on linux_amd64
```

### Additional information
Here is debug log with the sensitive content removed.
```
2022-11-25T07:36:25.494Z [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/hashicorp/kubernetes/2.16.0/linux_amd64/terraform-provider-kubernetes_v2.16.0_x5
2022-11-25T07:36:25.586Z [INFO]  provider.terraform-provider-kubernetes_v2.16.0_x5: configuring server automatic mTLS: timestamp=2022-11-25T07:36:25.586Z
2022-11-25T07:36:25.613Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: plugin address: address=/tmp/plugin151586155 network=unix timestamp=2022-11-25T07:36:25.613Z
2022-11-25T07:36:25.614Z [DEBUG] provider: using plugin: version=5
2022-11-25T07:36:25.652Z [DEBUG] created provider logger: level=debug
2022-11-25T07:36:25.652Z [INFO]  provider: configuring client automatic mTLS
2022-11-25T07:36:25.700Z [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/hashicorp/helm/2.7.1/linux_amd64/terraform-provider-helm_v2.7.1_x5 args=[.terraform/providers/registry.terraform.io/hashicorp/helm/2.7.1/linux_amd64/terraform-provider-helm_v2.7.1_x5]
2022-11-25T07:36:25.703Z [INFO]  provider.terraform-provider-google_v4.44.1_x5: 2022/11/25 07:36:25 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Content-Length: 760
Cache-Control: private, max-age=0, must-revalidate, no-transform
Content-Type: application/json; charset=UTF-8
Date: Fri, 25 Nov 2022 07:36:25 GMT
Etag: CMTo95Dpw/sCEAE=
Expires: Fri, 25 Nov 2022 07:36:25 GMT
Server: UploadServer
Vary: Origin
Vary: X-Origin
xxxxxx

-----------------------------------------------------: timestamp=2022-11-25T07:36:25.701Z
2022-11-25T07:36:25.703Z [INFO]  provider.terraform-provider-google_v4.44.1_x5: 2022/11/25 07:36:25 [DEBUG] Retry Transport: Stopping retries, last request was successful: timestamp=2022-11-25T07:36:25.702Z
2022-11-25T07:36:25.703Z [INFO]  provider.terraform-provider-google_v4.44.1_x5: 2022/11/25 07:36:25 [DEBUG] Retry Transport: Returning after 1 attempts: timestamp=2022-11-25T07:36:25.702Z
2022-11-25T07:36:25.703Z [WARN]  Provider "registry.terraform.io/hashicorp/google" produced an unexpected new value for xxxxxxxxxxxxx during refresh.
      - .media_link: was cty.StringVal("https://storage.googleapis.com/download/storage/v1/b/xxxxxx"), but now cty.StringVal("https://storage.googleapis.com/download/storage/v1/b/xxxxxxx")
2022-11-25T07:36:25.704Z [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/hashicorp/helm/2.7.1/linux_amd64/terraform-provider-helm_v2.7.1_x5 pid=564
2022-11-25T07:36:25.704Z [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/hashicorp/helm/2.7.1/linux_amd64/terraform-provider-helm_v2.7.1_x5
2022-11-25T07:36:25.746Z [INFO]  provider.terraform-provider-helm_v2.7.1_x5: configuring server automatic mTLS: timestamp=2022-11-25T07:36:25.746Z
2022-11-25T07:36:25.772Z [DEBUG] provider.terraform-provider-helm_v2.7.1_x5: plugin address: address=/tmp/plugin1646986889 network=unix timestamp=2022-11-25T07:36:25.772Z
2022-11-25T07:36:25.773Z [DEBUG] provider: using plugin: version=5
2022-11-25T07:36:25.868Z [WARN]  Provider "registry.terraform.io/hashicorp/google" produced an invalid plan for google_storage_bucket_object.updated_subnet_list, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .temporary_hold: planned value cty.False for a non-computed attribute
      - .content_encoding: planned value cty.StringVal("") for a non-computed attribute
      - .content_language: planned value cty.StringVal("") for a non-computed attribute
      - .event_based_hold: planned value cty.False for a non-computed attribute
      - .content_disposition: planned value cty.StringVal("") for a non-computed attribute
      - .detect_md5hash: planned value cty.StringVal("/sQe3fkiLscm8VzdL3vdRQ==") for a non-computed attribute
      - .metadata: planned value cty.MapValEmpty(cty.String) for a non-computed attribute
      - .cache_control: planned value cty.StringVal("") for a non-computed attribute
2022-11-25T07:36:25.878Z [DEBUG] provider.terraform-provider-google_v4.44.1_x5: 2022/11/25 07:36:25 [DEBUG] [transport] transport: http2Server.HandleStreams failed to read frame: read unix /tmp/plugin3491920865->@: use of closed network connection 
2022-11-25T07:36:25.878Z [DEBUG] provider.terraform-provider-google_v4.44.1_x5: 2022/11/25 07:36:25 [DEBUG] [transport] transport: error closing conn during Close: use of closed network connection 
2022-11-25T07:36:25.878Z [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/google/4.44.1/linux_amd64/terraform-provider-google_v4.44.1_x5 pid=532
2022-11-25T07:36:25.878Z [DEBUG] provider: plugin exited
2022-11-25T07:36:25.878Z [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2022-11-25T07:36:25.878Z [WARN]  ValidateProviderConfig from "provider[\"registry.terraform.io/hashicorp/helm\"]" changed the config value, but that value is unused
2022-11-25T07:36:25.881Z [INFO]  provider.terraform-provider-helm_v2.7.1_x5: 2022/11/25 07:36:25 [DEBUG] Experiments enabled: []: timestamp=2022-11-25T07:36:25.880Z
2022-11-25T07:36:26.088Z [WARN]  ValidateProviderConfig from "provider[\"registry.terraform.io/hashicorp/kubernetes\"]" changed the config value, but that value is unused
2022-11-25T07:36:26.529Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: 2022/11/25 07:36:26 [WARN] Invalid provider configuration was supplied. Provider operations likely to fail: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
2022-11-25T07:36:26.529Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: 2022/11/25 07:36:26 [DEBUG] Enabling HTTP requests/responses tracing
2022-11-25T07:36:26.529Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: 2022-11-25T07:36:26.529Z [ERROR] [Configure]: Failed to load config:="&{0xc000f734a0 0xc000169680 <nil> 0xc00018d2c0 {0 0} 0xc000ed58a0}"
2022-11-25T07:36:26.539Z [DEBUG] ReferenceTransformer: "module.kubernetes.data.google_client_config.default" references: []
2022-11-25T07:36:26.541Z [INFO]  ReferenceTransformer: reference not found: "module.gke"
2022-11-25T07:36:26.541Z [INFO]  ReferenceTransformer: reference not found: "var.enable_pd_ssd_storage_class"
2022-11-25T07:36:26.541Z [DEBUG] ReferenceTransformer: "kubernetes_storage_class.pd_ssd_storage_class[0]" references: []
2022-11-25T07:36:26.544Z [INFO]  ReferenceTransformer: reference not found: "module.gke"
2022-11-25T07:36:26.544Z [INFO]  ReferenceTransformer: reference not found: "var.install_cert_manager"
2022-11-25T07:36:26.544Z [INFO]  ReferenceTransformer: reference not found: "var.project_id"
2022-11-25T07:36:26.544Z [INFO]  ReferenceTransformer: reference not found: "local.project_google_service_account"
2022-11-25T07:36:26.544Z [INFO]  ReferenceTransformer: reference not found: "local.identity_namespace"
2022-11-25T07:36:26.544Z [INFO]  ReferenceTransformer: reference not found: "local.tools_namespace"
2022-11-25T07:36:26.544Z [DEBUG] ReferenceTransformer: "google_service_account_iam_member.cert_manager[0]" references: []
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "module.gke"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.identity_namespace"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.tools_namespace"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.identity_namespace"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.tools_namespace"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.gitlab_sa_name"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.identity_namespace"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.tools_namespace"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.external_secrets_release_name"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.identity_namespace"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.tools_namespace"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.gitlab_sa_name"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "var.project_id"
2022-11-25T07:36:26.546Z [INFO]  ReferenceTransformer: reference not found: "local.project_google_service_account"
2022-11-25T07:36:26.546Z [DEBUG] ReferenceTransformer: "google_service_account_iam_binding.gitlab_account_iam" references: []
2022-11-25T07:36:26.547Z [INFO]  ReferenceTransformer: reference not found: "module.gke"
2022-11-25T07:36:26.547Z [INFO]  ReferenceTransformer: reference not found: "local.gke_service_account"
2022-11-25T07:36:26.547Z [DEBUG] ReferenceTransformer: "google_service_account_key.gke_sa_key" references: []
2022-11-25T07:36:26.555Z [INFO]  ReferenceTransformer: reference not found: "module.gke"
2022-11-25T07:36:26.555Z [INFO]  ReferenceTransformer: reference not found: "var.install_cert_manager"
2022-11-25T07:36:26.555Z [INFO]  ReferenceTransformer: reference not found: "local.project_google_service_account"
2022-11-25T07:36:26.555Z [DEBUG] ReferenceTransformer: "google_service_account_key.project_sa_key[0]" references: []
2022-11-25T07:36:26.556Z [INFO]  ReferenceTransformer: reference not found: "module.gke"
2022-11-25T07:36:26.556Z [INFO]  ReferenceTransformer: reference not found: "var.create_gitlab_runner_cache_bucket"
2022-11-25T07:36:26.556Z [INFO]  ReferenceTransformer: reference not found: "local.project_name"
2022-11-25T07:36:26.556Z [INFO]  ReferenceTransformer: reference not found: "var.environment"
2022-11-25T07:36:26.556Z [INFO]  ReferenceTransformer: reference not found: "var.region"
2022-11-25T07:36:26.556Z [INFO]  ReferenceTransformer: reference not found: "var.project_id"
2022-11-25T07:36:26.556Z [INFO]  ReferenceTransformer: reference not found: "var.region"
2022-11-25T07:36:26.556Z [DEBUG] ReferenceTransformer: "google_storage_bucket.gitlab_cache_bucket[0]" references: []
2022-11-25T07:36:26.557Z [INFO]  ReferenceTransformer: reference not found: "module.gke"
2022-11-25T07:36:26.557Z [INFO]  ReferenceTransformer: reference not found: "local.tools_namespace"
2022-11-25T07:36:26.557Z [INFO]  ReferenceTransformer: reference not found: "local.tools_namespace"
2022-11-25T07:36:26.557Z [DEBUG] ReferenceTransformer: "kubernetes_namespace.eng_tools" references: []
2022-11-25T07:36:26.558Z [INFO]  provider.terraform-provider-google_v4.44.1_x5: 2022/11/25 07:36:26 [INFO] Instantiating Google Cloud IAM client for path https://iam.googleapis.com/: timestamp=2022-11-25T07:36:26.554Z
2022-11-25T07:36:26.559Z [INFO]  provider.terraform-provider-google_v4.44.1_x5: 2022/11/25 07:36:26 [DEBUG] Locking "iam-service-account-projects/xxxxxxxxxxxxxxxxxxxxx": timestamp=2022-11-25T07:36:26.555Z
2022-11-25T07:36:26.559Z [INFO]  provider.terraform-provider-google_v4.44.1_x5: 2022/11/25 07:36:26 [DEBUG] Locked "iam-service-account-projects/xxxxxxxxxxxxxxxxx": timestamp=2022-11-25T07:36:26.555Z
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: 2022/11/25 07:36:26 [INFO] Checking storage class pd-ssd
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: 2022/11/25 07:36:26 [DEBUG] Kubernetes API Request Details:
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: ---[ REQUEST ]---------------------------------------
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: GET /apis/storage.k8s.io/v1/storageclasses/pd-ssd HTTP/1.1
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: Host: localhost
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: User-Agent: HashiCorp/1.0 Terraform/1.0.9
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: Accept: application/json, */*
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: Accept-Encoding: gzip
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: 
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: 
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: -----------------------------------------------------
2022-11-25T07:36:26.570Z [DEBUG] provider.terraform-provider-kubernetes_v2.16.0_x5: 2022/11/25 07:36:26 [DEBUG] Received error: &url.Error{Op:"Get", URL:"http://localhost/apis/storage.k8s.io/v1/storageclasses/pd-ssd", Err:(*net.OpError)(0xc000f81ea0)}
2022-11-25T07:36:26.571Z [ERROR] provider.terraform-provider-kubernetes_v2.16.0_x5: Response contains error diagnostic: tf_proto_version=5.3 diagnostic_severity=ERROR @module=sdk.proto diagnostic_detail= diagnostic_summary="Get "http://localhost/apis/storage.k8s.io/v1/storageclasses/pd-ssd": dial tcp 127.0.0.1:80: connect: connection refused" tf_provider_addr=registry.terraform.io/hashicorp/kubernetes tf_req_id=b1265434-435f-d548-8a93-6629bce60c5f @caller=github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/internal/diag/diagnostics.go:55 tf_resource_type=kubernetes_storage_class tf_rpc=ReadResource timestamp=2022-11-25T07:36:26.568Z
```

**Comment by bharathkkb** [comment]
Thanks for the detailed report @FuyuanChu! I suspect this might be because of the existing k8s resources in your statefile trying to refresh but since the end point is only know after apply it is unable to use the endpoint. Could you try a targeted apply as a workaround or disable refresh? After migration when then endpoint is created, this should work normally.

**Comment by bharathkkb** [comment]
@FuyuanChu feel free to reopen if still an issue.
